### PR TITLE
Don't allocate node cidrs when using CNI

### PR DESCRIPTION
**Pull request by bboreham**
**What this PR does / why we need it**:

Absent this PR, the `kube-up.sh` configuration on GCE and AWS is set to always allocate a CIDR to each node and set up routes for those CIDRs.  This is unnecessary when using a CNI plugin, as it should be 
controlling all required network configuration.  Also, the routes can interfere with the operation of the CNI network, should it happen to use the same address range.

Once we stop the cloud provider setting up routes, on GCE nothing gets scheduled because kubelet sets the state to `NetworkUnavailable` and expects this to be cleared when the route is added.  This behaviour is noted at #27347, #26983 and #27828.  To avoid that problem, we add a plugin capability `NET_PLUGIN_CAPABILITY_CLOUD_BASED` which only `kubenet` implements; thereby if kubelet is using a CNI plugin (or an `exec` plugin), it will not set `NetworkUnavailable`.

I might go one step further and remove the gce-only test in `pkg/kubelet/kubelet.go`; it seems it was added in a hurry as a result of #27347.  If we can be confident that all third-party networks are working via a plugin then this PR will cover the requirement and allow #27828 to be marked fixed. cc @freehan

**Release note**:

``` release-note
When using a CNI network plugin, don't allocate node cidrs, create cloud routes or set the node state to NetworkUnavailable until a route is added.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33438)

### CODE REVIEW

**Comment by bboreham** [comment]
Rebased to remove the stray commit which caused the "authored by someone else" message

File: pkg/kubelet/kubelet_network.go [hunk 1]
```diff
@@ -124,6 +124,9 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
 	if kl.cloud == nil || kl.cloud.ProviderName() != "gce" || kl.flannelExperimentalOverlay {
 		return false
 	}
+	if kl.networkPlugin != nil && !kl.networkPlugin.Capabilities().Has(network.NET_PLUGIN_CAPABILITY_CLOUD_BASED) {
```
Comment (justinsb) [inline pkg/kubelet/kubelet_network.go]: Propose naming this `NET_PLUGIN_CAPABILITY_NODE_READINESS`, and we should glog.Warningf if the network provider is not validating that nodes are ready before allowing them to be scheduled.  It seems this is particularly important with NetworkPolicy!

File: cluster/aws/util.sh [hunk 2]
```diff
@@ -36,7 +36,9 @@ source "${KUBE_ROOT}/cluster/aws/${KUBE_CONFIG_FILE-"config-default.sh"}"
 source "${KUBE_ROOT}/cluster/common.sh"
 source "${KUBE_ROOT}/cluster/lib/util.sh"
 
-ALLOCATE_NODE_CIDRS=true
+if [[ "${NETWORK_PROVIDER:-}" != "cni" ]]; then
```
Comment (justinsb) [inline cluster/aws/util.sh]: Some CNI plugins _will_ use the k8s API though.  Doesn't flannel do this today?  My kopeio-routing definitely does (though I'm still leveraging kubenet!)

**Review by justinsb — CHANGES_REQUESTED** [review]
In general, I think node network readiness is an important part of a correct network provider, and I don't think we should be encouraging providers to opt out...  I guess if we put a warning that you're running a deprecated network provider or something similar?

File: pkg/kubelet/kubelet_network.go (hunk 1 above)
Comment (justinsb) [inline pkg/kubelet/kubelet_network.go]: Actually looks like NODE_AVAILABLE is more in tune with the naming of the condition.  I just think "cloud" is going to cause confusion :-)

File: cluster/aws/util.sh (hunk 2 above)
Comment (bboreham) [inline cluster/aws/util.sh]: There is no mention of such a dependency at https://github.com/containernetworking/cni/blob/master/Documentation/flannel.md

CNI is used by more than just Kubernetes; I guess it's possible that someone has built a Kubernetes-specific CNI plugin, but I'm prepared to worry about that when I see it.

File: pkg/kubelet/kubelet_network.go (hunk 1 above)
Comment (bboreham) [inline pkg/kubelet/kubelet_network.go]: What did you mean by "the network provider is not validating that nodes are ready"?

The current behaviour is that code inside kubelet marks the node as not-ready, and code inside the controller-manager then marks it as ready when it thinks it ought to be, without anyone validating anything.

**Comment by bboreham** [comment]
> node network readiness is an important part of a correct network provider

I agree.  Unfortunately there is no "ready" handshake with CNI at the moment.  That could be the subject of a separate PR.

>  put a warning that you're running a deprecated network provider

I can't figure out which providers you think should be deprecated.  Everything that is not kubenet?

File: pkg/kubelet/kubelet_network.go (hunk 1 above)
Comment (bboreham) [inline pkg/kubelet/kubelet_network.go]: renamed to `NET_PLUGIN_CAPABILITY_NODE_AVAILABLE`

File: pkg/kubelet/network/plugins.go [hunk 3]
```diff
@@ -45,7 +45,8 @@ const NET_PLUGIN_EVENT_POD_CIDR_CHANGE_DETAIL_CIDR = "pod-cidr"
 // Plugin capabilities
 const (
 	// Indicates the plugin handles Kubernetes bandwidth shaping annotations internally
-	NET_PLUGIN_CAPABILITY_SHAPING int = 1
+	NET_PLUGIN_CAPABILITY_SHAPING        int = 1
+	NET_PLUGIN_CAPABILITY_NODE_AVAILABLE int = 2
```
Comment (dcbw) [inline pkg/kubelet/network/plugins.go]: Need some godoc here for what the new condition is for.

**Comment by dcbw** [comment]
@bboreham for the cloud routes thing, see https://github.com/kubernetes/kubernetes/pull/34398 for how I proposed we should solve it.

**Comment by caseydavenport** [comment]
> This is unnecessary when using a CNI plugin, as it should be controlling all required network configuration

Maybe I'm missing something, but this isn't actually true.  It's totally reasonable to have a CNI plugin that relies on using cloud routes for getting traffic between hosts.  This is exactly how kubenet works (and the intention is still to move kubenet into its own CNI plugin).

bridge CNI + host-local IPAM + cloud routes is perfectly valid.

File: cluster/aws/util.sh (hunk 2 above)
Comment (caseydavenport) [inline cluster/aws/util.sh]: It doesn't need to be a Kubernetes specific CNI plugin - bridge + host-local is an example that would rely on cloud routes being configured.

Flannel does have an integration with Kubernetes where it expects the podCIDR field to be populated, and uses CNI.  See [kube-flannel](https://github.com/coreos/flannel/blob/master/Documentation/kube-flannel.yml) + [code](https://github.com/coreos/flannel/blob/5dde68dbb9df5318768ab68d42fd055a94547f88/subnet/kube/kube.go).  It won't strictly need the cloud routes themselves, but does depend on the node cidrs being allocated.

As mentioned in another comment, Calico CNI also has an option to operate in a mode where it expects the podCIDR to exist and cloud routes to be configured.

**Review by caseydavenport — CHANGES_REQUESTED** [review]

**Comment by dcbw** [comment]
All these things (cloud routes, master allocated CIDRs, network plugin availability) can/should be independently selected because some plugins use one and some use others, some use all of them.  Yeah, that makes it harder to autoconfigure a cluster, but that's life...

**Comment by bboreham** [comment]
I prefer the way #34398 controls internal state.  That PR does not touch the original intent of this one, which was to stop setting unnecessary cloud routes, however I think `kube-up.sh` is becoming less relevant so going to close this.

**Pull request closed without merging**